staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,482 @@
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares or
+// we are at max queue length
+// 4) If not rejected, create a packet and enqueue
+// returns true on a successful enqueue
+// returns false in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+func (fqs *FQScheduler) TimeoutOldRequestsAndRejectOrEnqueue(hashValue uint64, handSize int32) FQPacket {
+	fqs.lock.Lock()
+	defer fqs.lock.Unlock()
+
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := fqs.ChooseQueueIdx(hashValue, int(handSize))
+	queue := fqs.GetQueues()[queueIdx]
+	// The next step is the logic to reject requests that have been waiting too long
+	fqs.removeTimedOutPacketsFromQueue(queue)
+	// NOTE: currently timeout is only checked for each new request.  This means that there can be
+	// requests that are in the queue longer than the timeout if there are no new requests
+	// We think this is a fine tradeoff
+
+	// Create a packet and enqueue
+	pkt := FQPacket(&Packet{
+		DequeueChannel: make(chan bool, 1),
+		EnqueueTime:    time.Now(),
+		Queue:          queue,
+	})
+	if ok := fqs.rejectOrEnqueue(pkt); !ok {
+		return nil
+	}
+	return pkt
+
+}
+
+// removeTimedOutPacketsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (fqs *FQScheduler) removeTimedOutPacketsFromQueue(queue FQQueue) {
+	timeoutIdx := -1
+	now := time.Now()
+	pkts := queue.GetPackets()
+	// pkts are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+	for i, pkt := range pkts {
+		channelPkt := pkt.(*Packet)
+		limit := channelPkt.EnqueueTime.Add(fqs.requestWaitLimit)
+		if now.After(limit) {
+			channelPkt.DequeueChannel <- false
+			close(channelPkt.DequeueChannel)
+			// // TODO(aaron-prindle) verify this makes sense here
+			// get idx for timed out packets
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	// TODO(aaron-prindle) BAD/IMPROVE, currently copies slice for deletion
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.SetPackets(pkts[removeIdx:])
+		fqs.DecrementPackets(removeIdx)
+	}
+}
+
+// DecrementPackets decreases the # of packets for the FQScheduler w/ lock
+func (fqs *FQScheduler) DecrementPackets(i int) {
+	fqs.numPackets -= i
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishPacket method invoked after service)
+func (fqs *FQScheduler) getRequestsExecuting() int {
+	total := 0
+	for _, queue := range fqs.Queues {
+		total += queue.GetRequestsExecuting()
```

Comment:
This could cost O(1) instead of O(NumQueues) if this sum were maintained incrementally.